README.md
# License Plate Detection
This project is done as a Final Year Project required to complete a Bachelor's degree in Electronics Engineering. The code is based on Adrian Rosebrock's (@jrosebr1) PyImageSearch post [**OpenCV: Automatic License/Number Plate Recognition (ANPR) with Python**](https://www.pyimagesearch.com/2020/09/21/opencv-automatic-license-number-plate-recognition-anpr-with-python/), edited to be able to detect Malaysian license plates which typically has white texts on black background plate. I recommend reading through his blogpost to get a general idea of how the image processing pipeline works.

The project aims to:
* Automatically detects either Malaysian license plates or other oversea license plates
* Compare Scharr operator, Canny edge detector, and edge-less approach to detect license plates

## Prerequisites
The project code utilizes the following library:
* [Python](https://www.python.org/) v3.8.6
* [OpenCV](https://opencv.org/) v4.4.0.46
* [imutils](https://github.com/jrosebr1/imutils) v0.5.3
* [scikit-image](https://scikit-image.org/) v0.17.2
* [pytesseract](https://github.com/madmaze/pytesseract) v0.3.6
* [Tesseract-OCR](https://tesseract-ocr.github.io/tessdoc/) v5.0.0 (alpha). (This is NOT installed through pip but with the Windows installer)

## How it works
1. The project has 2 modules which is the driver and the class module named **anprdriver.py** and **anprclass.py** respectively. The driver module contains the script which the user should execute through Python while providing the arguments whereas the class module contains the three detection algorithm (Sobel, Canny, and edge-less) and other functions including saving results and workflow debug display.
2. The driver should be executed in a terminal by using Python (`$ python anprdriver.py ...`). The arguments which could be appended are (arguments in bold are required):
   - **`-i` (input)** - path to input directory of images
   - `-c` (clear border) - should the script clear border pixels before performing OCR (default is -1, set to 1 to enable)
   - `-p` (PSM) - default tesseract page segmentation mode for license plate OCR (default is 7)
   - `-d` (debug) - should additional visualizations be shown (default is -1, set to 1 to enable)
   - `-a` (algorithm) - specify which algorithm to use (default is 1, 1 for Scharr, 2 for Canny, 3 for edge-less)
   - `-s` (save) - whether to save the results or not (default is -1, *disable debug* and set to 1 to enable)
   - `-m` (morphology) - specify morphological operation (default is 1, 1 for black hat which is better for black text on white plates, 2 for top hat which is better for the opposite)
3. The repository includes sample images of license plates in `malaysian` and `overseas` folder. Results are saved inside `result_canny_*`, `result_sobel_*`, and `result_edgeless_*` folders respectively where `*` is either `malaysian` or `overseas`.

## Typical usage
The script can be typically run by typing `python anprdriver.py -i /path/to/images/`. Path to the images can be written as `-i ./malaysian` for example. While only -i argument is necessary, be sure to choose the correct algorithm (-a) and morphological operation (-m) as needed.

## README TO BE UPDATED FURTHER


CHANGELOG.md
# Changelog

## Added

- Comments have been added where applicable to better clarify the functionality.
- Folders to save the final resulting images will now be created automatically and named according to the input image folder name.

## Changed

- Absolute path for saving the final resulting images had been changed to relative path. Note that the `pytesseract.pytesseract.tesseract_cmd` still uses absolute path.

## Removed

- Removed processing time measurement and recording as I deemed it to be unreliable to measure actual image processing pipeline performance.
- Removed redundant `locate_license_plate()` functions inside `CannyANPR` and `EdgelessANPR` class in anprclass.py.

## Fixed

- Fixed an issue where the program still saves final resulting images despite debug (-d) being enabled.


anprdriver.py
from anprclass import CannyANPR, EdgelessANPR, SobelANPR
from imutils import paths
import argparse
import imutils
import sys
import cv2
import os

test = os.getcwd()
print(test)

def cleanup_text(text):
	return "".join([c if ord(c) < 128 else "" for c in text]).strip()

ap = argparse.ArgumentParser()
ap.add_argument("-i", "--input", required=True,
	help="Path to input directory of images")
ap.add_argument("-c", "--clear-border", type=int, default=-1,
	help="Whether or to clear border pixels before OCR'ing")
ap.add_argument("-p", "--psm", type=int, default=7,
	help="Default PSM mode for OCR'ing license plates")
ap.add_argument("-d", "--debug", type=int, default=-1,
	help="Whether or not to show additional visualizations")
ap.add_argument("-a", "--algorithm", type=int, default=1,
    help="Choose an edge detection method (1 - Sobel edge detection, 2 - Canny edge detection, 3 - Edge-less approach")
ap.add_argument("-s", "--save", type=int, default = -1,
    help="Whether to save or not the results in a folder")
ap.add_argument("-m", "--morphology", type=str, default = 'bh',
    help="Whether to use black hat (-m bh) or top hat (-m th). Black hat is better at detecting black-on-white license plate while top hat is the opposite.")
args = vars(ap.parse_args())

anpr = None
iteration = 0

input_dir = args["input"]
input_dir_name = input_dir.split('/')[-1]

algo = args["algorithm"]
if algo == 1:
    anpr = SobelANPR(algo, input_dir_name, morph=args["morphology"], debug=args["debug"] > 0, save=args["save"] > 0)
elif algo == 2:
    anpr = CannyANPR(algo, input_dir_name, morph=args["morphology"], debug=args["debug"] > 0, save=args["save"] > 0)
elif algo == 3:
    anpr = EdgelessANPR(algo, input_dir_name, morph=args["morphology"], debug=args["debug"] > 0, save=args["save"] > 0)
else:
    print('Invalid algorithm choice')
    sys.exit()

imagePaths = sorted(list(paths.list_images(input_dir)))
for imagePath in imagePaths:
    iteration += 1
    originimage = cv2.imread(imagePath)
    image = imutils.resize(originimage, width=400, height=400)

    image = cv2.bilateralFilter(image, 3, 105, 105)
    anpr.debug_imshow("Bilateral Filter", image, waitKey=True)

    (lpText, lpCnt) = anpr.find_and_ocr(iteration, image, psm=args["psm"], clearBorder=args["clear_border"] > 0)

    if lpText is not None and lpCnt is not None:
        box = cv2.boxPoints(cv2.minAreaRect(lpCnt))
        box = box.astype("int")
        cv2.drawContours(image, [box], -1, (0, 255, 0), 2)

        (x, y, w, h) = cv2.boundingRect(lpCnt)
        cv2.putText(image, cleanup_text(lpText), (x, y - 15),
            cv2.FONT_HERSHEY_SIMPLEX, 0.75, (0, 255, 0), 2)

        print("[INFO] Registration number: {}".format(lpText))
        anpr.debug_imshow("Output ANPR", image, waitKey=True)

    anpr.save_result("Final{}.jpg".format(iteration), image)
    cv2.destroyAllWindows()

anprclass.py
from skimage.segmentation import clear_border
import pytesseract
import numpy as np
import imutils
import cv2
import os

#Path to Tesseract executable. Usually located at C:\Program Files\Tesseract-OCR\tesseract.exe for Windows.
#This is only required if Tesseract executable is not defined in PATH, otherwise keep the line below commented.
pytesseract.pytesseract.tesseract_cmd = r'C:\\Users\\afeef\\Documents\\Development\\TesseractOCR\\tesseract.exe' #CHANGE THIS BEFORE COMMIT

class SobelANPR:
    def __init__(self, algo, input_dir, morph, minAR=2.5, maxAR=5, debug=False, save=False):  
        self.minAR = minAR  #These two lines are to define the aspect ratio (AR) of the region of interest (ROI). Try experimenting with these values!
        self.maxAR = maxAR  #If the region of interest is within the boundary (minAR and maxAR) set here, it will be considered as a license plate
        self.debug = debug
        self.save = save
        self.algo = algo
        self.input_dir = input_dir
        self.morph = morph

    def debug_imshow(self, title, image, waitKey=False):    #If debug argument (-d) is set to 1, the script will show the whole image processing pipeline
        if self.debug:                                      #and wait for user input before continuing to the next step.
            cv2.imshow(title, image)
            if waitKey:
                cv2.waitKey(0)

    def save_result(self, name, image):                     #This function is used to save the final image which contains the ROI
        try:                                                #in a folder and also the image of the extracted ROI.
            result_dir = ['result_sobel', 'result_canny', 'result_edgeless']    
            image_dir = r'.' + '\{}'.format(result_dir[self.algo-1] + '_{}'.format(self.input_dir))
            print(image_dir)
            os.chdir(image_dir)
        except:
            print("No folder directory found. Creating...")
            os.mkdir(image_dir)
            os.chdir(image_dir)
            
        if self.save and self.debug != 1:
            cv2.imwrite(name, image)
            
        os.chdir('..')

    def morphology_operation(self, gray, rectKern):
        if self.morph=='bh':
            blackhat = cv2.morphologyEx(gray, cv2.MORPH_BLACKHAT, rectKern)
            self.debug_imshow("Blackhat", blackhat, waitKey=True)

            squareKern = cv2.getStructuringElement(cv2.MORPH_RECT, (3, 3))
            light = cv2.morphologyEx(gray, cv2.MORPH_CLOSE, squareKern)
            self.debug_imshow("Closing operation", light, waitKey=True)
            light = cv2.threshold(light, 0, 255,
                cv2.THRESH_BINARY | cv2.THRESH_OTSU)[1]
            self.debug_imshow("Light Regions", light, waitKey=True)

            return [blackhat, light]

        elif self.morph=='th':
            tophat = cv2.morphologyEx(gray, cv2.MORPH_TOPHAT, rectKern)
            self.debug_imshow("Tophat", tophat, waitKey=True)

            squareKern = cv2.getStructuringElement(cv2.MORPH_RECT, (3, 3))
            dark = cv2.morphologyEx(gray, cv2.MORPH_CLOSE, squareKern)
            self.debug_imshow("Closing operation", dark, waitKey=True)
            dark = cv2.threshold(dark, 0, 255,
                cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)[1]
            self.debug_imshow("Dark Regions", dark, waitKey=True)

            return [tophat, dark]

    def locate_license_plate_candidates(self, gray, image, keep=5):
        rectKern = cv2.getStructuringElement(cv2.MORPH_RECT, (13, 5))
        morphology = self.morphology_operation(gray, rectKern)
        morph = morphology[0]
        luminance = morphology[1]

        gradX = cv2.Sobel(morph, ddepth=cv2.CV_32F,
            dx=1, dy=0, ksize=3)
        gradX = np.absolute(gradX)
        (minVal, maxVal) = (np.min(gradX), np.max(gradX))
        gradX = 255 * ((gradX - minVal) / (maxVal - minVal))
        gradX = gradX.astype("uint8")
        self.debug_imshow("Scharr", gradX, waitKey=True)

        gradX = cv2.GaussianBlur(gradX, (5, 5), 0)
        self.debug_imshow("Gaussian", gradX, waitKey=True)
        gradX = cv2.morphologyEx(gradX, cv2.MORPH_CLOSE, rectKern)
        thresh = cv2.threshold(gradX, 0, 255,
            cv2.THRESH_BINARY | cv2.THRESH_OTSU)[1]
        self.debug_imshow("Grad Thresh", thresh, waitKey=True)

        thresh = cv2.erode(thresh, None, iterations=3)
        thresh = cv2.dilate(thresh, None, iterations=3)
        self.debug_imshow("Grad Erode/Dilate", thresh, waitKey=True)

        thresh = cv2.bitwise_and(thresh, thresh, mask=luminance)
        thresh = cv2.dilate(thresh, None, iterations=2)
        thresh = cv2.erode(thresh, None, iterations=1)

        cnts = cv2.findContours(thresh.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        cnts = imutils.grab_contours(cnts)
        cnts = sorted(cnts, key=cv2.contourArea, reverse=True)[:keep]

        oriCopy = image.copy()
        for c in cnts:
            cv2.drawContours(oriCopy, [c], -1, 255, 2)
            self.debug_imshow("Contours", oriCopy)

        self.debug_imshow("Masked", thresh, waitKey=True)

        return cnts

    def locate_license_plate(self, iteration, gray, candidates, clearBorder=False):
        lpCnt = None
        roi = None

        candidates = sorted(candidates, key=cv2.contourArea)

        for c in candidates:
            (x, y, w, h) = cv2.boundingRect(c)
            ar = w / float(h)

            if ar >= self.minAR and ar <= self.maxAR:
                lpCnt = c
                licensePlate = gray[y:y + h, x:x + w]
                roi = cv2.threshold(licensePlate, 0, 255, cv2.THRESH_BINARY_INV | cv2.THRESH_OTSU)[1]

                if clearBorder:
                    roi = clear_border(roi)

                self.debug_imshow("License Plate", licensePlate, waitKey=True)
                self.debug_imshow("ROI", roi, waitKey=True)

                self.save_result('roi{}.png'.format(iteration), licensePlate)
                break

        return (roi, lpCnt)

    def build_tesseract_options(self, psm=7):
        alphanumeric = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        options = "-c tessedit_char_whitelist={}".format(alphanumeric)
        options += " --psm {}".format(psm)
        return options

    def find_and_ocr(self, iteration, image, psm=7, clearBorder=False):
        lpText = None

        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        self.debug_imshow("Grayscale", gray, waitKey=True)
        candidates = self.locate_license_plate_candidates(gray, image, 5)
        (lp, lpCnt) = self.locate_license_plate(iteration, gray, candidates, clearBorder=clearBorder)

        if lp is not None:
            options = self.build_tesseract_options(psm=psm)
            lpText = pytesseract.image_to_string(lp, config=options)
            self.debug_imshow("License Plate", lp, waitKey=True)

        return (lpText, lpCnt)

class CannyANPR(SobelANPR):
    def locate_license_plate_candidates(self, gray, image, keep = 5):
        rectKern = cv2.getStructuringElement(cv2.MORPH_RECT, (13, 5))
        morphology = self.morphology_operation(gray, rectKern)
        morph = morphology[0]
        luminance = morphology[1]

        canny = cv2.Canny(morph, 200, 230) # Originally 400,450; try to experiment with these values.
        self.debug_imshow("Canny", canny, waitKey=True)

        gaussian = cv2.GaussianBlur(canny, (5,5), 0)
        gaussian = cv2.morphologyEx(gaussian, cv2.MORPH_CLOSE, rectKern)
        thresh = cv2.threshold(gaussian, 0, 255, 
            cv2.THRESH_BINARY | cv2.THRESH_OTSU)[1]
        self.debug_imshow("Grad Thresh", thresh, waitKey=True)
        
        thresh = cv2.erode(thresh, None, iterations=3)
        thresh = cv2.dilate(thresh, None, iterations=3)
        self.debug_imshow("Eroded & Dilated", thresh, waitKey=True)

        thresh = cv2.bitwise_and(thresh, thresh, mask=luminance)
        thresh = cv2.dilate(thresh, None, iterations=2)
        thresh = cv2.erode(thresh, None, iterations=1)
        
        cnts = cv2.findContours(thresh.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        cnts = imutils.grab_contours(cnts)
        cnts = sorted(cnts, key=cv2.contourArea, reverse=True)[:keep]

        oriCopy = image.copy()
        for c in cnts:
            cv2.drawContours(oriCopy, [c], -1, 255, 2)
            self.debug_imshow("Contours", oriCopy)

        self.debug_imshow("Masked", thresh, waitKey=True)

        return cnts

class EdgelessANPR(SobelANPR):
    def locate_license_plate_candidates(self, gray, image, keep = 5):
        rectKern = cv2.getStructuringElement(cv2.MORPH_RECT, (13, 5))
        morphology = self.morphology_operation(gray, rectKern)
        morph = morphology[0]
        luminance = morphology[1]

        gaussian = cv2.GaussianBlur(morph, (5,5), 0)
        thresh = cv2.morphologyEx(gaussian, cv2.MORPH_CLOSE, rectKern)
        thresh = cv2.threshold(thresh, 0, 255, 
            cv2.THRESH_BINARY | cv2.THRESH_OTSU)[1]
        self.debug_imshow("Thresholded", thresh, waitKey=True)

        thresh = cv2.erode(thresh, None, iterations=3)
        thresh = cv2.dilate(thresh, None, iterations=3)
        self.debug_imshow("Eroded & Dilated", thresh, waitKey=True)

        thresh = cv2.bitwise_and(thresh, thresh, mask=luminance)
        thresh = cv2.dilate(thresh, None, iterations=2)
        thresh = cv2.erode(thresh, None, iterations=1)
        
        cnts = cv2.findContours(thresh.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        cnts = imutils.grab_contours(cnts)
        cnts = sorted(cnts, key=cv2.contourArea, reverse=True)[:keep]

        oriCopy = image.copy()
        for c in cnts:
            cv2.drawContours(oriCopy, [c], -1, 255, 2)
            self.debug_imshow("Contours", oriCopy)

        self.debug_imshow("Masked", thresh, waitKey=True)

        return cnts